README.md
## Chill [![Build Status](https://secure.travis-ci.org/twitter/chill.png)](http://travis-ci.org/twitter/chill)

Extensions for the [Kryo serialization library](https://github.com/EsotericSoftware/kryo) including
serializers and a set of classes to ease configuration of Kryo in systems like Hadoop, Storm,
Akka, etc.


### Building Chill

```bash
./sbt
> compile # to build chill
> publishM2 # to publish chill to your local .m2 repo
> publish-local # publish to local ivy repo.
```

Chill has a set of subprojects: chill-java, chill-hadoop, chill-storm and chill-scala.  Other than
chill-scala, all these projects are written in Java so they are easy to use on any JVM platform.

## Chill-Java

The chill-java package includes the `KryoInstantiator` class (factory for Kryo instances)
and the `IKryoRegistrar` interface (adds Serializers to a given Kryo). These two are composable
to build instantiators that create instances of Kryo that have the options and serializers you
need. The benefit of this over a direct Kryo instance is that a Kryo instance is mutable and not
serializable, which limits the safety and reusability of code that works directly with them.

To deserialize or serialize easily, look at `KryoPool`:

```java
int POOL_SIZE = 10;
KryoPool kryo = KryoPool.withByteArrayOutputStream(POOL_SIZE, new KryoInstantiator());
byte[] ser = kryo.toBytesWithClass(myObj);
Object deserObj = kryo.fromBytes(myObj);
```

The KryoPool is a thread-safe way to share Kryo instances and temporary output buffers.

### Chill Config

Hadoop, Storm, and Akka all use a configuration that is basically equivalent to a `Map[String,
String]`. The `com.twitter.chill.config` package makes it easy to build up `KryoInstantiator`
instances given a Config instance, which is an abstract class acting as a thin wrapper over
whatever configuration data the system, such as Hadoop, Storm or Akka, might give.

To configure a KryoInstantiator use `ConfiguredInstantiator` with either reflection,
which takes a class name and instantiates that KryoInstantiator, or an instance of KryoInstantiator
and serializes that instance to use later:
```scala
class TestInst extends KryoInstantiator { override def newKryo = sys.error("blow up") }

// A new Config:
val conf = new JavaMapConfig
// Set-up class-based reflection of our instantiator:
ConfiguredInstantiator.setReflect(conf, classOf[TestInst])
val cci = new ConfiguredInstantiator(conf)
cci.newKryo // uses TestInst
//Or serialize a particular instance into the config to use later (or another node):

ConfiguredInstantiator.setSerialized(conf, new TestInst)
val cci2 = new ConfiguredInstantiator(conf)
cci2.newKryo // uses the particular instance we passed above
```

## Chill in Scala

Scala classes often have a number of properties that distinguish them from usual Java classes. Often
scala classes are immutable, and thus have no zero argument constructor. Secondly, `object` in scala is
a singleton that needs to be carefully serialized. Additionally, scala classes often have synthetic
(compiler generated) fields that need to be serialized, and by default Kryo does not serialize
those.

In addition to a `ScalaKryoInstantiator` which generates Kryo instances with options suitable for
scala, chill provides a number of Kryo serializers for standard scala classes (see below).

### The MeatLocker

Many existing systems use Java serialization. MeatLocker is an object that wraps a given instance
using Kryo serialization internally, but the MeatLocker itself is Java serializable.
The MeatLocker allows you to box Kryo-serializable objects and deserialize them lazily on the first call to `get`:

```scala
import com.twitter.chill.MeatLocker

val boxedItem = MeatLocker(someItem)

// boxedItem is java.io.Serializable no matter what it contains.
val box = roundTripThroughJava(boxedItem)
box.get == boxedItem.get // true!
```

To retrieve the boxed item without caching the deserialized value, use `meatlockerInstance.copy`.

### Serializers for Scala classes

These are found in the `chill-scala` directory in the chill jar (originally this project was
only scala serializers).  Chill provides support for singletons, scala Objects and the following types:

* Scala primitives
  * scala.Enumeration values
  * scala.Symbol
  * scala.reflect.Manifest
  * scala.reflect.ClassManifest
  * scala.Function[0-22] closure cleaning (removing unused `$outer` references).
* Collections and sequences
  * scala.collection.immutable.Map
  * scala.collection.immutable.List
  * scala.collection.immutable.Vector
  * scala.collection.immutable.Set
  * scala.collection.mutable.{Map, Set, Buffer, WrappedArray}
  * all 22 scala tuples

## Chill-bijection

[Bijections and Injections](https://github.com/twitter/bijection) are useful when considering serialization. If you have an Injection from `T` to `Array[Byte]` you have a serialization.  Additionally, if you have a Bijection between `A` and `B`, and a serialization for `B`, then you have a serialization for `A`.  See `BijectionEnrichedKryo` for easy interop between bijection and chill.

### KryoInjection: easy serialization to byte Arrays

KryoInjection is an injection from `Any` to `Array[Byte]`. To serialize using it:

```scala
import com.twitter.chill.KryoInjection

val bytes:  Array[Byte]    = KryoInjection(someItem)
val tryDecode: scala.util.Try[Any] = KryoInjection.invert(bytes)
```

KryoInjection can be composed with Bijections and Injections from `com.twitter.bijection`.

## Chill-Akka

To use, add a key to your config like:
```
    akka.actor.serializers {
      kryo = "com.twitter.chill.akka.AkkaSerializer"
    }
```

Then for the super-classes of all your message types, for instance, scala.Product, write:
```scala
   akka.actor.serialization-bindings {
     "scala.Product" = kryo
   }
```

If you want to use the `chill.config.ConfiguredInstantiator` see `ConfiguredAkkaSerializer`
otherwise, subclass `AkkaSerializer` and override `kryoInstantiator` to control how the `Kryo`
object is created.

## Community and Documentation

This, and all [github.com/twitter](https://github.com/twitter) projects, are under the [Twitter Open Source Code of Conduct](https://engineering.twitter.com/opensource/code-of-conduct). Additionally, see the [Typelevel Code of Conduct](http://typelevel.org/conduct) for specific examples of harassing behavior that are not tolerated.

To learn more and find links to tutorials and information around the web, check out the [Chill Wiki](https://github.com/twitter/chill/wiki).

The latest ScalaDocs are hosted on Chill's [Github Project Page](http://twitter.github.io/chill).

Discussion occurs primarily on the [Chill mailing list](https://groups.google.com/forum/#!forum/chill-user). Issues should be reported on the [GitHub issue tracker](https://github.com/twitter/chill/issues).

## Maven

Chill modules are available on Maven Central. The current groupid and version for all modules is, respectively, `"com.twitter"` and  `0.8.0` and each scala project is published for `2.10` and `2.11`. Search [search.maven.org](http://search.maven.org/#search%7Cga%7C1%7Cchill) when in doubt.

## Authors

* Oscar Boykin <https://twitter.com/posco>
* Mike Gagnon <https://twitter.com/MichaelNGagnon>
* Sam Ritchie <https://twitter.com/sritchie>

## License

Copyright 2012 Twitter, Inc.

Licensed under the Apache License, Version 2.0: http://www.apache.org/licenses/LICENSE-2.0


CHANGES.md
# chill #

### 0.8.0 ###
* Kryo 3.0.3 upgrade: https://github.com/twitter/chill/pull/245

### 0.7.4 ###
* If the inner Instantiator is cached we should just cache the kayo and…: https://github.com/twitter/chill/pull/243

### 0.7.3 ###
* Update the build: https://github.com/twitter/chill/pull/248
* Use currentThread classloader: https://github.com/twitter/chill/pull/247

### 0.7.2 ###
* Setting the class loader by default: https://github.com/twitter/chill/pull/242

### 0.7.1 ###
* Add testing support for Java 6, 7 and 8.
* Adds two Injection's to lift T into an Externalizer of T.
* Protobuf module is Java only.

### 0.7.0 ###
* Adds Scrooge 2.11 support now that the artifact is out, sets up for c…: https://github.com/twitter/chill/pull/236
* make chill 1.6 compatible: https://github.com/twitter/chill/pull/231
* Update README.md: https://github.com/twitter/chill/pull/232
* Alter scala collection default serializer registration order: https://github.com/twitter/chill/pull/229
* Build chill-scrooge for scala 2.11, too: https://github.com/twitter/chill/pull/219

### 0.6.0 ###
* Add build instructions to readme and make InjectiveSerializer serializable #216
* Build chill-scrooge for scala 2.11, too #219
* Rewrite Java BitSet serializer to make it more efficient #220
* Bijection 0.8.0, algebird 0.10.0, scala 2.10.5 #228

### 0.5.2 ###
* Use new Travis CI infrastructure #210
* Optimizations for ConfiguredInstantiator. #213
* Upgrade to algebird 0.9.0, bijection 0.7.2, scala 2.11.5 #214

### 0.5.1
* reverse Config contains logic: https://github.com/twitter/chill/pull/205
* fix setConf: https://github.com/twitter/chill/pull/204
* fix default constructor for kryo serialization: https://github.com/twitter/chill/pull/203
* Switched Chill Avro to use ClassTags, added Java unit tests: https://github.com/twitter/chill/pull/200
* Enable cross compilation for chill-avro: https://github.com/twitter/chill/pull/202

### 0.5.0
* Move to211: https://github.com/twitter/chill/pull/197
* Make 2.10.4 the default, move to scalatest: https://github.com/twitter/chill/pull/196

### 0.4.0
* Support serializing asJavaIterable.: https://github.com/twitter/chill/pull/192
* Remove a deprecation, that really should never have been there: https://github.com/twitter/chill/pull/189
* Use scalariform: https://github.com/twitter/chill/pull/188
* README: Update Kryo project URL.: https://github.com/twitter/chill/pull/187
* support mutable BitSet: https://github.com/twitter/chill/pull/185
* Make chill-avro work with generic records: https://github.com/twitter/chill/pull/184
* updating akka dependency to 2.3.2: https://github.com/twitter/chill/pull/182
* add chill-algebird project by copying AlgebirdSerializers from of scaldi...: https://github.com/twitter/chill/pull/177
* Scrooge serializer: https://github.com/twitter/chill/pull/178
* Use shaded asm classes provided by Kryo.: https://github.com/twitter/chill/pull/175

### 0.3.6
* Add ScalaAnyRefMapConfig, deals with non-string keys in cascading: https://github.com/twitter/chill/pull/174
* added AvroSerializer: https://github.com/twitter/chill/pull/172
* Add sbt script for homogeneity with SB, Scalding, etc: https://github.com/twitter/chill/pull/170
* Support for akka 2.2.1: https://github.com/twitter/chill/pull/169
* Protobuf should be in the all target: https://github.com/twitter/chill/pull/168
* Enable getClass equality checking in tests: https://github.com/twitter/chill/pull/165
* Add chill-protobuf, with tests: https://github.com/twitter/chill/pull/163
* support serialization of scala sortedmap: https://github.com/twitter/chill/pull/162

### 0.3.5
* Add Serializers for scala SortedList and ListSet: https://github.com/twitter/chill/pull/152
* Fix Range serialization (and remove broken subclass-serialization of Iterable, Seq, etc): https://github.com/twitter/chill/pull/154
* Build and test chill-akka for 2.10: https://github.com/twitter/chill/pull/155
* Add chill-thrift: https://github.com/twitter/chill/pull/156
* Support JavaConverter-built classes: https://github.com/twitter/chill/pull/159
* Back to 2.21: https://github.com/twitter/chill/pull/157
* Adds a test from issue #8: https://github.com/twitter/chill/pull/158

### 0.3.4
* Bugfixes for Externalizer with looped object graphs https://github.com/twitter/chill/pull/143

### 0.3.3
* Serialize synthetic fields by default: https://github.com/twitter/chill/pull/135
* Prefer Java to Kryo, but check both in Externalizer: https://github.com/twitter/chill/pull/138

### 0.3.2

* Add a LocaleSerializer to chill-java: https://github.com/twitter/chill/pull/128
* chill-akka module: https://github.com/twitter/chill/pull/116
* Community section to README (with a mailing list!) https://github.com/twitter/chill/pull/127
* SingletonSerializer for (): https://github.com/twitter/chill/pull/130
* Add Externalizer (fault-tolerant meatlocker): https://github.com/twitter/chill/pull/126
* Use References in Externalizer: https://github.com/twitter/chill/pull/131

### 0.3.1

* Issue 115: fix for wrapped Arrays with non-primitives

### 0.3.0
* Lots of refactoring around configuration (see chill.config in chill-java)
* Issue #94 KryoPool for pooling Kryo instances and Output buffers
* Issue #85 upgrade to bijection 0.5.2 and scala 2.9.3
* Issue #78 Some, Left, Right serializers in chill(-scala)
* Issue #70 create chill-java for Java only code
* Issue #67 add chill-storm
* Issue #65, #71 import cascading.kryo as chill-hadoop
* Issue #64 fix KryoInjection to be from Any to Array[Byte]
* Issues #61, #63 handle the case of small Map/Set in scala
* Issue #57 use CanBuildFrom in Traversable serialization

Contributors:

94 commits
* P. Oscar Boykin: 46 commits
* Sam Ritchie: 38 commits
* ryanlecompte: 5 commits
* Ngoc Dao: 2 commits
* Dao Ngoc: 1 commits
* Michael Schmitz: 1 commits
* Julien Le Dem: 1 commits

### 0.2.3
* Update to Kryo 2.21
* Update to Bijection 0.4.0
* Improve Traversable Serializer

### 0.2.2

* Custom RegexSerializer

project/Build.scala
package chill

import sbt._
import Keys._
import com.typesafe.tools.mima.plugin.MimaPlugin.mimaDefaultSettings
import com.typesafe.tools.mima.plugin.MimaKeys._
import scalariform.formatter.preferences._
import com.typesafe.sbt.SbtScalariform._

import scala.collection.JavaConverters._

object ChillBuild extends Build {
  val kryoVersion = "2.21"

  val bijectionVersion = "0.8.1"
  val algebirdVersion = "0.11.0"

  def isScala210x(scalaVersion: String) = scalaVersion match {
      case version if version startsWith "2.10" => true
      case _ => false
  }

  val sharedSettings = Project.defaultSettings ++ mimaDefaultSettings ++ scalariformSettings ++ Seq(
    organization := "com.twitter",
    scalaVersion := "2.10.5",
    crossScalaVersions := Seq("2.10.5", "2.11.7"),
    scalacOptions ++= Seq("-unchecked", "-deprecation"),
    ScalariformKeys.preferences := formattingPreferences,

    // Twitter Hadoop needs this, sorry 1.7 fans
    javacOptions ++= Seq("-target", "1.6", "-source", "1.6", "-Xlint:-options"),
    javacOptions in doc := Seq("-source", "1.6"),

    resolvers ++= Seq(
      Opts.resolver.sonatypeSnapshots,
      Opts.resolver.sonatypeReleases
    ),
    libraryDependencies ++= Seq(
      "org.scalacheck" %% "scalacheck" % "1.12.2" % "test",
      "org.scalatest" %% "scalatest" % "2.2.4" % "test",
      "com.esotericsoftware.kryo" % "kryo" % kryoVersion
    ),

    parallelExecution in Test := true,

    // Publishing options:
    publishMavenStyle := true,
    publishTo <<= version { v =>
      Some(
        if (v.trim.toUpperCase.endsWith("SNAPSHOT"))
          Opts.resolver.sonatypeSnapshots
        else
          Opts.resolver.sonatypeStaging
          //"twttr" at "http://artifactory.local.twitter.com/libs-releases-local"
      )
    },
    publishArtifact in Test := false,
    pomIncludeRepository := { x => false },
    pomExtra := (
      <url>https://github.com/twitter/chill</url>
          <licenses>
        <license>
        <name>Apache 2</name>
        <url>http://www.apache.org/licenses/LICENSE-2.0.txt</url>
          <distribution>repo</distribution>
        <comments>A business-friendly OSS license</comments>
        </license>
        </licenses>
        <scm>
        <url>git@github.com:twitter/chill.git</url>
        <connection>scm:git:git@github.com:twitter/chill.git</connection>
        </scm>
        <developers>
        <developer>
        <id>oscar</id>
        <name>Oscar Boykin</name>
        <url>http://twitter.com/posco</url>
          </developer>
        <developer>
        <id>sritchie</id>
        <name>Sam Ritchie</name>
        <url>http://twitter.com/sritchie</url>
          </developer>
        </developers>)
  )

  // Aggregated project
  lazy val chillAll = Project(
    id = "chill-all",
    base = file("."),
    settings = sharedSettings
  ).settings(
    test := { },
    publish := { },
    publishLocal := { }
  ).aggregate(
    chill,
    chillBijection,
    chillScrooge,
    chillStorm,
    chillJava,
    chillHadoop,
    chillThrift,
    chillProtobuf,
    chillAkka,
    chillAvro,
    chillAlgebird
    )

  lazy val formattingPreferences = {
   import scalariform.formatter.preferences._
   FormattingPreferences().
     setPreference(AlignParameters, false).
     setPreference(PreserveSpaceBeforeArguments, true)
  }

  /**
    * This returns the youngest jar we released that is compatible
    * with the current.
    */
  val unreleasedModules = Set[String]("akka")
  val javaOnly = Set[String]("storm", "java", "hadoop", "thrift", "protobuf")

  def youngestForwardCompatible(subProj: String) =
    Some(subProj)
      .filterNot(unreleasedModules.contains(_))
      .map { s =>
      val suffix = if (javaOnly.contains(s)) "" else "_2.10"
      "com.twitter" % ("chill-" + s + suffix) % "0.7.1"
    }

  def module(name: String) = {
    val id = "chill-%s".format(name)
    Project(id = id, base = file(id), settings = sharedSettings ++ Seq(
      Keys.name := id,
      previousArtifact := youngestForwardCompatible(name),
      // Disable cross publishing for java artifacts
      publishArtifact <<= (scalaVersion) { scalaVersion =>
        if(javaOnly.contains(name) && scalaVersion.startsWith("2.11")) false else true
      }
      )
    )
  }

  // We usually do the pattern of having a core module, but we don't want to cause
  // pain for legacy deploys. With this, they can stay the same.
  lazy val chill = Project(
    id = "chill",
    base = file("chill-scala"),
    settings = sharedSettings
  ).settings(
    name := "chill",
    previousArtifact := Some("com.twitter" % "chill_2.10" % "0.7.0")
  ).dependsOn(chillJava)

  lazy val chillAkka = module("akka").settings(
    resolvers += "Typesafe Repository" at "http://repo.typesafe.com/typesafe/releases/",
    libraryDependencies ++= Seq(
      "com.typesafe" % "config" % "1.2.1",
      "com.typesafe.akka" %% "akka-actor" % "2.3.6" % "provided"
    )
  ).dependsOn(chill % "test->test;compile->compile")

  lazy val chillBijection = module("bijection").settings(
    libraryDependencies ++= Seq(
      "com.twitter" %% "bijection-core" % "0.8.0"
    )
  ).dependsOn(chill % "test->test;compile->compile")

  // This can only have java deps!
  lazy val chillJava = module("java").settings(
    crossPaths := false,
    autoScalaLibrary := false
  )

  // This can only have java deps!
  lazy val chillStorm = module("storm").settings(
    crossPaths := false,
    autoScalaLibrary := false,
    resolvers ++= Seq(
      "Clojars Repository" at "http://clojars.org/repo",
      "Conjars Repository" at "http://conjars.org/repo"
    ),
    libraryDependencies += "storm" % "storm" % "0.9.0-wip9" % "provided"
  ).dependsOn(chillJava)

  // This can only have java deps!
  lazy val chillHadoop = module("hadoop").settings(
    crossPaths := false,
    autoScalaLibrary := false,
    libraryDependencies ++= Seq(
      "org.apache.hadoop" % "hadoop-core" % "0.20.2" % "provided",
      "org.slf4j" % "slf4j-api" % "1.6.6",
      "org.slf4j" % "slf4j-log4j12" % "1.6.6" % "provided"
    )
  ).dependsOn(chillJava)

  // This can only have java deps!
  lazy val chillThrift = module("thrift").settings(
    crossPaths := false,
    autoScalaLibrary := false,
    libraryDependencies ++= Seq(
      "org.apache.thrift" % "libthrift" % "0.6.1" % "provided"
    )
  )

  lazy val chillScrooge = module("scrooge").settings(
    libraryDependencies ++= Seq(
      "org.apache.thrift" % "libthrift" % "0.6.1" exclude("junit", "junit"),
      "com.twitter" %% "scrooge-serializer" % "3.20.0"
    )
  ).dependsOn(chill % "test->test;compile->compile")

  // This can only have java deps!
  lazy val chillProtobuf = module("protobuf").settings(
    crossPaths := false,
    autoScalaLibrary := false,
    libraryDependencies ++= Seq(
      "com.google.protobuf" % "protobuf-java" % "2.3.0" % "provided"
    )
  ).dependsOn(chillJava)

  lazy val chillAvro = module("avro").settings(
    libraryDependencies ++= Seq(
      "com.twitter" %% "bijection-avro" % "0.8.0",
      "junit" % "junit" % "4.5" % "test"
    )
  ).dependsOn(chill,chillJava, chillBijection)

  lazy val chillAlgebird = module("algebird").settings(
    libraryDependencies ++= Seq(
      "com.twitter" %% "algebird-core" % "0.10.0"
    )
  ).dependsOn(chill)
}
